compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -676,9 +676,61 @@ pub(super) fn implied_predicates_with_filter<'tcx>(
         _ => {}
     }
 
+    assert_only_contains_predicates_from(filter, implied_bounds, tcx.types.self_param);
+
     ty::EarlyBinder::bind(implied_bounds)
 }
 
+// Make sure when elaborating supertraits, probing for associated types, etc.,
+// we really truly are elaborating clauses that have `Self` as their self type.
+// This is very important since downstream code relies on this being correct.
+pub(super) fn assert_only_contains_predicates_from<'tcx>(
+    filter: PredicateFilter,
+    bounds: &'tcx [(ty::Clause<'tcx>, Span)],
+    ty: Ty<'tcx>,
+) {
+    if cfg!(debug_assertions) {
```

Comment:
doesn't this trigger lint unused-variables for the params if not(debug_assertions)? we can prob just prefix them with underscores like we do in similar assertion fns.

personally I would go for an early return to avoid the rightward drift but doesn't matter

Reply:
No, because `if cfg!(debug_assertions)` expands to `if false` and it still gets used

Reply:
ah true, I must've thought of `#[cfg(...)] { ... }`

Reply:
Oh, yeah, that's dumb. IDK why I wrote it like this. Probably because I only pulled it out of a method after I wrote the `if`.